src/inmanta/agent/cache.py

```diff
@@ -307,6 +311,69 @@ def get_or_else(
+            for_version: Optional[bool], timeout: Optional[int], evict_after_last_access: float, evict_after_creation: float
+        ) -> tuple[float, float]:
+            """
+            This method is a backwards compatibility layer to compute a "new-style" retention policy (i.e. that is using
+            evict_after_last_access and/or evict_after_creation semantics) from the parameters of the `get_or_else` method.
+
+            :param for_version: Compatibility rules for this deprecated parameter:
+                If passed and True, entries will expire <evict_after_last_access>s after their last access (60s default)
+                If passed and False, entries will expire a certain number of seconds after entering the cache. This time
+                is either <evict_after_creation>, <timeout> or a default of 5000s, whichever is set, inspected in this order.
+            :param timeout: Compatibility rules for this deprecated parameter:
+                If `for_version=False` and the new-style parameter `evict_after_creation` is not set, this parameter
+                controls the expiry time (in seconds) of entries after entering the cache.
+                If `for_version` is not set, this parameter is an alias for the new-style parameter `evict_after_creation`.
+                (If both are set, the new-style parameter `evict_after_creation` has precedence)
+            :param evict_after_last_access: This cache item will be considered stale this number of seconds after
+                it was last accessed.
+            :param evict_after_creation: This cache item will be considered stale this number of seconds after
+                entering the cache.
+            """
+            _evict_after_last_access: float
+            _evict_after_creation: float
+
+            # Legacy `for_version` parameter is used, compute
+            # evict_after_last_access and evict_after_creation
+            if for_version is not None:
+                if for_version:
+                    _evict_after_creation = 0
+                    _evict_after_last_access = evict_after_last_access if evict_after_last_access > 0 else 60
+                else:
+                    _evict_after_last_access = 0
+                    if evict_after_creation > 0:
+                        _evict_after_creation = evict_after_creation
+                    elif timeout and timeout > 0:
+                        _evict_after_creation = timeout
+                    else:
+                        _evict_after_creation = 5000
+
+                    if evict_after_creation > 0 and timeout and timeout > 0:
+                        LOGGER.warning(
+                            "Both the `evict_after_creation` and the deprecated `timeout` parameter are set "
+                            "for cached method %s. Cached entries will be kept in the cache for %.2fs "
+                            "after entering it.",
+                            key,
```

Comment:
I'm not super satisfied with the solution I came up with:
I considered the bare `key` argument (i.e. without the `kwargs` prepended to it) to be the method name.
- This works fine in the context of the `@cache` decorator since the handler passes the mehod name `key=f.__name__` when calling into `get_or_else`
- In the context of calling the `get_or_else` on its own, this warning will look weird since the `key` argument can be anything i.e. not necessarily a method name

Reply:
I will actually log the cache entry key instead.

Reply:
Logging the entry key doesn't make any sense. The standard way the cache should be used it through the `@cache` annotation. In that case the key is chosen by our internal logic, which the user doesn't even know about. The user only knows that the method will return a cached entry if available. Using the method name would be a lot better. It's a unique identifier. At least, I don't see a reason why you would put two different cache policies on the same method.

Reply:
There is also a problem with putting this warning inside the `get_or_else()` method. It means that the warning will be logged on every call to the method with the cache attached. So if the method is used a lot, it will result in a very big amount of log messages filling up the disk. If we move the warning to the `cache()` method (the annotation), only one warning will be logged when the annotation is executed and None when the cache is used.

Reply:
I would also make the log message more actionable. We should clearly explain to the user what should be done to resolve this warning. Right now this is quite implicit.